### DISCUSSION TITLE
How to add an include path to a vscode project?

### DISCUSSION THREAD

**Discussion by datascribbler777** [microsoft/vscode-cpptools#10088] [Q&A] [answered]
Asking this here because I can't seem to find an answer anywhere else. Really trying to give this IDE a shot, but this is a pretty basic example and it is going nowhere.

I need to see a working example of how to add an include path to vscode.

My directory structure:

C:.
├───.vscode
│       c_cpp_properties.json
│       launch.json
│       tasks.json
│
├───inc
│       funcs.h
│
└───src
        funcs.cpp
        main.cpp

funcs.h:
```
#pragma once

void sayHello();
```

funcs.cpp:
```
#include <iostream>
#include "funcs.h"

void sayHello()
{
    std::cout << "Hello!" << std::endl;
}
```

main.cpp:
```
#include "funcs.h"

int main()
{
    sayHello();
    return 0;
}
```

VSCode marks red squigglies under the #include "funcs.h" in main.cpp, but not in funcs.cpp for some reason. VSCode says it cannot find funcs.h. If I try to build the compiler throws the same error.

I have tried adding an include path for the inc folder by editing the C/C++ Configuration page and in the c_cpp_properties.json file. By default it already contains ${workspaceFolder}/** which is supposed to search recursively through all the directories. All of the googling I have tried says to add the path here but never gives the exact syntax.

I have tried adding:
/inc
./inc
../inc
/inc/.h
${workspaceFolder}/inc
${workspaceFolder}/inc/.h
${workspaceFolder}/inc/funcs.h
and more combinations that I would care to list here.

How do I make vscode see my .h file in my inc folder?

**Comment by bobbrow** [comment]
Can you run the `C/C++: Log Diagnostics` command when the file with the red squiggles is open in the editor and paste the result here?

**Reply by datascribbler777** [reply]
```
-------- Diagnostics - 10/31/2022, 4:33:10 PM
Version: 1.12.4
Current Configuration:
{
    "name": "Win32",
    "includePath": [
        "${workspaceFolder}/**",
        "MSVSC_hell/inc"
    ],
    "defines": [
        "_DEBUG",
        "UNICODE",
        "_UNICODE"
    ],
    "compilerPath": "C:/mingw64/bin/g++.exe",
    "cStandard": "c17",
    "cppStandard": "c++17",
    "intelliSenseMode": "windows-gcc-x64",
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": true,
    "mergeConfigurations": false,
    "browse": {
        "path": [
            "${workspaceFolder}/**",
            "MSVSC_hell/inc"
        ],
        "limitSymbolsToIncludedHeaders": true
    }
}
Translation Unit Mappings:
[ C:\Users\PEL6FH\Desktop\temp\MSVSC_hell\src\funcs.cpp ]:
    C:\Users\PEL6FH\Desktop\temp\MSVSC_hell\src\funcs.cpp
[ C:\Users\PEL6FH\Desktop\temp\MSVSC_hell\src\main.cpp ]:
    C:\Users\PEL6FH\Desktop\temp\MSVSC_hell\src\main.cpp
Translation Unit Configurations:
[ C:\Users\PEL6FH\Desktop\temp\MSVSC_hell\src\funcs.cpp ]:
    Process ID: 13972
    Memory Usage: 63 MB
    Compiler Path: C:/mingw64/bin/g++.exe
    Includes:
        C:\MINGW64\LIB\GCC\X86_64-W64-MINGW32\12.2.0\INCLUDE\C++
        C:\MINGW64\LIB\GCC\X86_64-W64-MINGW32\12.2.0\INCLUDE\C++\X86_64-W64-MINGW32
        C:\MINGW64\LIB\GCC\X86_64-W64-MINGW32\12.2.0\INCLUDE\C++\BACKWARD
        C:\MINGW64\LIB\GCC\X86_64-W64-MINGW32\12.2.0\INCLUDE
        C:\MINGW64\LIB\GCC\X86_64-W64-MINGW32\12.2.0\INCLUDE-FIXED
        C:\MINGW64\X86_64-W64-MINGW32\INCLUDE
        C:\USERS\PEL6FH\DESKTOP\TEMP\MSVSC_HELL\INC
    Defines:
        _DEBUG
        UNICODE
        _UNICODE
    Standard Version: c++17
    IntelliSense Mode: windows-gcc-x64
    Other Flags:
        --g++
        --gnu_version=120200
[ C:\Users\PEL6FH\Desktop\temp\MSVSC_hell\src\main.cpp ]:
    Process ID: 2284
    Memory Usage: 13 MB
    Compiler Path: C:/mingw64/bin/g++.exe
    Includes:
        C:\MINGW64\LIB\GCC\X86_64-W64-MINGW32\12.2.0\INCLUDE\C++
        C:\MINGW64\LIB\GCC\X86_64-W64-MINGW32\12.2.0\INCLUDE\C++\X86_64-W64-MINGW32
        C:\MINGW64\LIB\GCC\X86_64-W64-MINGW32\12.2.0\INCLUDE\C++\BACKWARD
        C:\MINGW64\LIB\GCC\X86_64-W64-MINGW32\12.2.0\INCLUDE
        C:\MINGW64\LIB\GCC\X86_64-W64-MINGW32\12.2.0\INCLUDE-FIXED
        C:\MINGW64\X86_64-W64-MINGW32\INCLUDE
        C:\USERS\PEL6FH\DESKTOP\TEMP\MSVSC_HELL\INC
    Defines:
        _DEBUG
        UNICODE
        _UNICODE
    Standard Version: c++17
    IntelliSense Mode: windows-gcc-x64
    Other Flags:
        --g++
        --gnu_version=120200
Total Memory Usage: 76 MB

------- Workspace parsing diagnostics -------
Number of files discovered (not excluded): 2555
```

I actually set `${workspaceFolderBasename}/inc` for my include path, but it evaluates to the path listed in the diags, which is the correct location. 
If I don't add anything and leave the `${workspaceFolder}/**`,  it basically says it can't find my header file. If I put in `${workspaceFolder}/inc` it looks in some other path that is not in my project. Maybe I am not understanding what `${workspaceFolder}` is supposed to represent.

And thank you for the response.

**Reply by bobbrow** [reply]
`${workspaceFolder}` maps to the folder currently open in VS Code.  Is `C:\Users\PEL6FH\Desktop\temp\MSVSC_hell` the currently open folder?  If so, I would have expected both `${workspaceFolder}/**` and `${workspaceFolder}/inc` to work.

**Reply by bobbrow** [reply]
Another idea... Are you using any filesystem tricks like symlinks, or case-sensitive folders in this path?

**Reply by datascribbler777** [reply]
I am using windows 10, no other "tricks" that I know of or are doing intentionally. 
When I put in just {workspaceFolder}/inc it was coming up with an error message about being unable to find some crazy long path. 
I was have been trying to get it to repeat that, but it is not doing that anymore, of course. 

That is why I started trying {workspaceFolderBasename}/inc, that seemed to evaluate to the correct name.

I will keep trying to get that error back.

**Reply by datascribbler777** [reply]
Another weird thing, when I try to build and look at the output it is not adding any -I switch for include paths.  When I go to the tasks.json file and add it myself, it ends up using it but puts double quotes around it and the compiler chokes. 

```
{
    "tasks": [
        {
            "type": "cppbuild",
            "label": "C/C++: g++.exe build active file",
            "command": "C:/mingw64/bin/g++.exe",
            "args": [
                "-fdiagnostics-color=always",
                "-g",
                "${file}",
                "-I inc",
                "-o",
                "${fileDirname}\\${fileBasenameNoExtension}.exe"
            ],
            "options": {
                "cwd": "C:/mingw64/bin"
            },
            "problemMatcher": [
                "$gcc"
            ],
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "detail": "Task generated by Debugger."
        }
    ],
    "version": "2.0.0"
}
```

results in:
`g++.exe -fdiagnostics-color=always -g C:\Users\PEL6FH\Desktop\temp\MSVSC_hell\src\main.cpp "-I inc" -o C:\Users\PEL6FH\Desktop\temp\MSVSC_hell\src\main.exe`

**Reply by bobbrow** [reply]
Each spaced item on the command line needs to be a separate item in the args array.  So use: `"-I", "inc"` instead of `"-I inc"`

**Reply by datascribbler777** [reply]
Ok, I tried that and it is now putting in the switch and relative path, but it still gives and error for no such file or directory for "funcs.h"

**Reply by datascribbler777** [reply]
Also just tried a full uninstall/reinstall, including the appdate/Code directory and the User/.vscode directory. I also created a new directory and re-typed out the code. Same error.

**Reply by bobbrow** [reply]
Oh, I think it's because the cwd is pointing to the compiler path (so that compilation works).  Try replacing `"inc"` with `"${workspaceFolder}/inc"`

**Reply by datascribbler777** [reply]
If I use "Show Errors" it says
func.h: No such file or directory gcc [Ln 1, Col10]

Does it matter that it says gcc when I am have set my compiler to g++?

**Reply by datascribbler777** [reply]
> Oh, I think it's because the cwd is pointing to the compiler path (so that compilation works). Try replacing `"inc"` with `"${workspaceFolder}/inc"`

In the tasks.json or in the include path settings?

**Reply by bobbrow** [reply]
tasks.json

**Comment by datascribbler777** [comment, marked as answer]
I added it to the tasks.json
The red squiggles are gone from main.cpp
It still fails, but now for a slightly different error:
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\PEL6FH\AppData\Local\Temp\ccrYS7OC.o:C:/Users/PEL6FH/Desktop/temp/MSVSC_hell/src/main.cpp:5: undefined reference to `sayHello()'
collect2.exe: error: ld returned 1 exit status

And the program started acting really strange. Mouseclicks are not being registered. Only one file can be open in the editor now, if I try to open a new one it closes the current one/tab and opens the new one in its place.

**Reply by bobbrow** [reply]
sayHello is declared in funcs.h, but defined in funcs.cpp.  You will need to build main.cpp and funcs.cpp for the linker to be able to find the implementation.

This Stack Overflow post may shed some light on what you need to do: https://stackoverflow.com/questions/18777326/compiling-multiple-c-files-with-gcc

I typically prefer to use CMake or Makefiles to manage compiling multiple files into a single executable binary, so I'm not sure if there's a better way than what I'm going to propose, but you should be able to do it by creating two tasks.  The first one compiles the two source files `main.cpp` and `funcs.cpp` into `.o` files.  Then the second task links the object files into a binary and adds the `"dependsOn"` property to point to the first task.

Something like this (not tested, may not work as written):
```json
{
    "tasks": [
        {
            "type": "cppbuild",
            "label": "compile",
            "command": "C:/mingw64/bin/g++.exe",
            "args": [
                "-fdiagnostics-color=always",
                "-g",
                "-I",
                "${workspaceFolder}/inc",
                "main.cpp",
                "funcs.cpp"
            ],
            "options": {
                "cwd": "C:/mingw64/bin"
            },
            "problemMatcher": [
                "$gcc"
            ]
        },
        {
            "type": "cppbuild",
            "label": "build",
            "command": "C:/mingw64/bin/g++.exe",
            "args": [
                "-g",
                "main.o",
                "funcs.o",
                "-o",
                "myBinary.exe"
            ],
            "options": {
                "cwd": "C:/mingw64/bin"
            },
            "problemMatcher": [
                "$gcc"
            ],
            "group": {
                "kind": "build",
                "isDefault": true
            },
           "dependsOn": [ "compile" ]
        }
    ],
    "version": "2.0.0"
}
```

**Reply by bobbrow** [reply]
Oops. I posted the example json with some obvious errors. They're fixed now.

**Reply by datascribbler777** [reply]
Ok, I was wondering how this would work in this tool. I may just do it by hand in the console, much easier for me to understand. I would usually do something like:
g++ src/main.cpp src/funcs.cpp -I inc -o main

I was hoping vscode would do something similar once I setup the settings.

**Reply by datascribbler777** [reply]
So the official answer to how to include an include path is it "should" work if you type it into the projects C/C++ Configurations.

But for me, that didn't work. So instead I had to add the include path using vscode variables, like ${workspaceFolder}<your_path> into the tasks.json file. 

I really appreciate the help Bob. 

I can't seem to figure out how to mark this as "answered."  If you or a mod can delete it, that might be better. It doesn't seem like anyone else is having the problems I am having. I am probably just going to go back to notepad++ and cmd. Editing .json's is not for me, especially if I will have to "figure out" what will need to be added where. In the future I will be using a much more developed directory structure and having to integrate other tools like gtest, conan, etc...

**Reply by bobbrow** [reply]
Glad I could be of help.  I'll mark this as the answer.  I definitely recommend learning/using a build system of some kind if your project is going to consist of more than a handful of source files.  CMake and Makefile are two choices that we support in VS Code and would help you avoid configuring json files.  CMake works well with gtest and conan/vcpkg and you should be able to find examples out there to guide you.
